src/pages/api/gateway/[gatewayUID]/sensor/[macAddress]/config.ts

```diff
@@ -1,95 +1,100 @@
+    res.json({ err: `Method ${req.method || "is undefined."} Not Allowed` });
+    return false;
+  }
+  return true;
+}
+function validateRequest(
   req: NextApiRequest,
   res: NextApiResponse
-) {
-  // Gateway UID must be a string
-  if (typeof req.query.gatewayUID !== "string") {
-    res.status(400).json({ err: HTTP_STATUS.INVALID_GATEWAY });
-    return;
+): false | ValidRequest {
+  const { gatewayUID, macAddress } = req.query;
+  // TODO (carl): figure out what to do about this unsafe assignment of any. I really want runtime
+  // typechecking. All the code below feels terrible to write when we've already specified the API
+  // in typescript. Maybe we should really use a different language to specify the API and generate
+  // the typescript and typechecking code?
+  const { name, location } = req.body;
+
+  if (typeof gatewayUID !== "string") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: HTTP_STATUS.INVALID_GATEWAY });
+    return false;
   }
-  // Sensor MAC must be a string
-  if (typeof req.query.macAddress !== "string") {
-    res.status(400).json({ err: HTTP_STATUS.INVALID_SENSOR_MAC });
-    return;
+  if (typeof macAddress !== "string") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: HTTP_STATUS.INVALID_SENSOR_MAC });
+    return false;
+  }
+  if (typeof location !== "string" && typeof location !== "undefined") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: "location should be a string or undefined" });
+    return false;
+  }
+  if (typeof name !== "string" && typeof name !== "undefined") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: "name should be a string or undefined" });
+    return false;
+  }
+  if (typeof location === "undefined" && typeof name === "undefined") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: HTTP_STATUS.INVALID_SENSOR_CONFIG_BODY });
+    return false;
   }
 
-  // Query params
-  const { gatewayUID, macAddress } = req.query;
-  // Body params
-  const { loc, name } = req.body as NoteSensorConfigBody;
-  // Notehub values
-  const { hubBaseURL, hubAuthToken, hubProjectUID } = config;
-  // API path
-  const endpoint = `${hubBaseURL}/req?project=${hubProjectUID}&device=${gatewayUID}`;
-  // API headers
-  const headers = {
-    [HTTP_HEADER.CONTENT_TYPE]: HTTP_HEADER.CONTENT_TYPE_JSON,
-    [HTTP_HEADER.SESSION_TOKEN]: hubAuthToken,
-  };
+  return { gatewayUID, macAddress, location, name };
+}
 
-  // note.update payload
-  const noteUpdate = {
-    req: "note.update",
-    file: "config.db",
-    note: macAddress,
-    body: {
-      loc,
-      name,
-    },
-  };
+async function performRequest({
+  macAddress,
+  gatewayUID,
+  location,
+  name,
+}: ValidRequest) {
+  const app = services().getAppService();
+  try {
+    if (location) await app.setSensorLocation(gatewayUID, macAddress, location);
+    if (name) await app.setSensorName(gatewayUID, macAddress, name);
```

Comment:
This is going to hit the notehub API 4 times when both location and name are set.  Just wondering if having each attribute as a separate setter is the right design here, especially as the list of attributes will undoubtedly grow. 

How about a `builder` pattern, where the attributes are set on a builder instance, e.g.
```
app.updateSensorAttributes(gatewayUID, macAddress)
.setLocation(location)
.setName(name)
.update();
``` 

The attributes can be fetched in `updateSensorAttributes()` and then set once in the call to `update()`.

Reply:
With the way the UI is (or at least will be) you won’t be able to update both at the same time. Not that we should necessarily let the UI dictate the backend APIs, but in practice I don’t think this will be performance problem.

Reply:
I'm with TJ on this for now.  If the time comes for multiple update we'll figure something out.